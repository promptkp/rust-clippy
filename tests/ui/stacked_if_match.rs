#![warn(clippy::stacked_if_match)]
#![allow(unused)]

fn stacked_if() {
    let x = 0;
    if if x == 1 { x == 2 } else { x == 3 } {
        println!("true");
    }

    if if x == 1 { 2 } else { 3 } == 4 {
        println!("true");
    }

    if if x == 1 {
        let y = 2;
        y == 2
    } else {
        let z = 3;
        z == 3
    } {
        println!("true");
    }
}

fn stacked_match() {
    let x = 0;
    match match x {
        1 => 2,
        _ => 3,
    } {
        1 => {},
        2 => {},
        _ => {},
    }

    match match x {
        1 => 2,
        _ => 3,
    } + 1
    {
        1 => {},
        2 => {},
        _ => {},
    }
}

fn if_no_lint() {
    let x = 0;

    if (if x == 1 { x == 2 } else { x == 3 }) {
        println!("true");
    }

    if 1 == if x == 1 { 1 } else { 2 } {
        println!("true");
    }
}

fn match_no_lint() {
    let x = 0;
    match (match x {
        1 => 2,
        _ => 3,
    }) {
        1 => {},
        2 => {},
        _ => {},
    }

    match 1 + match x {
        1 => 2,
        _ => 3,
    } {
        1 => {},
        2 => {},
        _ => {},
    }
}

fn if_match_no_lint() {
    let x = 0;
    if match x {
        1 => 2,
        _ => 3,
    } == 4 {
        println!("true");
    }
}

fn match_if_no_lint() {
    let x = 0;
    match if x == 1 {
        let y = 2;
        y + 1
    } else {
        let z = 3;
        z + 1
    } {
        1 => {},
        2 => {},
        _ => {},
    }
}

macro_rules! if_macro {
    ($var:ident) => {
        if $var == 1 { true } else { false }
    };
}

macro_rules! match_macro {
    ($var:ident) => {
        match $var {
            1 => 2,
            _ => 3,
        }
    };
}

macro_rules! if_if_macro {
    () => {
        if if true { true } else { false } {
            println!("true");
        }
    };
}

fn macro_no_lint() {
    let x = 0;
    if if_macro!(x) {
        println!("true");
    }

    match match_macro!(x) {
        1 => {},
        2 => {},
        _ => {},
    }

    if_if_macro!();
}

fn main() {}
