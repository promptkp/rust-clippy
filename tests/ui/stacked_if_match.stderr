error: avoid using `if if` by binding inner `if` with `let`
  --> tests/ui/stacked_if_match.rs:6:5
   |
LL |     if if x == 1 { x == 2 } else { x == 3 } {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `let result = if x == 1 { x == 2 } else { x == 3 }; if result`
   |
   = note: `-D clippy::stacked-if-match` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::stacked_if_match)]`

error: avoid using `if if` by binding inner `if` with `let`
  --> tests/ui/stacked_if_match.rs:10:5
   |
LL |     if if x == 1 { 2 } else { 3 } == 4 {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `let result = if x == 1 { 2 } else { 3 }; if result`

error: avoid using `if if` by binding inner `if` with `let`
  --> tests/ui/stacked_if_match.rs:14:5
   |
LL | /     if if x == 1 {
LL | |         let y = 2;
LL | |         y == 2
LL | |     } else {
LL | |         let z = 3;
LL | |         z == 3
LL | |     } {
   | |_____^
   |
help: try
   |
LL ~     let result = if x == 1 {
LL +         let y = 2;
LL +         y == 2
LL +     } else {
LL +         let z = 3;
LL +         z == 3
LL ~     }; if result {
   |

error: avoid using `match match` by binding inner `match` with `let`
  --> tests/ui/stacked_if_match.rs:27:5
   |
LL | /     match match x {
LL | |         1 => 2,
LL | |         _ => 3,
LL | |     } {
   | |_____^
   |
help: try
   |
LL ~     let result = match x {
LL +         1 => 2,
LL +         _ => 3,
LL ~     }; match result {
   |

error: avoid using `match match` by binding inner `match` with `let`
  --> tests/ui/stacked_if_match.rs:36:5
   |
LL | /     match match x {
LL | |         1 => 2,
LL | |         _ => 3,
LL | |     } + 1
   | |_____^
   |
help: try
   |
LL ~     let result = match x {
LL +         1 => 2,
LL +         _ => 3,
LL ~     }; match result + 1
   |

error: aborting due to 5 previous errors

